aips/aip-order-votes.md

```diff
@@ -0,0 +1,226 @@
+- The signature on the above ledger info.
+
+```
+pub struct OrderVote {
+    /// The identity of the voter.
+    author: Author,
+    /// LedgerInfo of a block that is going to be ordered in case this vote gathers QC.
+    ledger_info: LedgerInfo,
+    /// Signature of the LedgerInfo
+    signature: bls12381::Signature,
+}
+```
+
+In most cases, a validator first receives the quorum certificate on the block `r` before receiving order votes on the block. But when a validator’s network connection is faulty, the validator could potentially fail to receive the quorum certificate on block `r` but receive the order votes on the block `r`. In this case, without having the quorum certificate on block `r`, even if the validator receives enough order votes on block `r`, the validator will still have to wait for block `r+1` to execute the block `r`. To simplify this, the validators additionally append the quorum certificate on block `r` when broadcasting the order vote.
+
+```
+pub struct OrderVoteMsg {
+	order_vote: OrderVote
+    // The quorum certificate on the block that is going to be ordered
+	quorum_cert: QuorumCert
+}
+```
+
+When a validator obtains the QC on a block `r`, the validator uses the following algorithm to create the corresponding order vote.
+
+```
+Procedure process_vote {
+	// same as before
+	...
+	if a QC is aggregated {
+		order_vote <- make_order_vote(QC)
+		order_vote_msg <- OrderVoteMsg {order_vote, QC}
+		broadcast order_vote_msg
+	}
+}
+
+Procedure make_order_vote(qc) {
+	// Safety need to update the highest qc here
+	observe_qc(qc)
+	if safe_to_order_vote(qc) {
+		ledger_info <- LedgerInfo { qc.vote_data.proposed, HashValue::dummy() }
+		signature <- sign(ledger_info)
+		order_vote <- OrderVote { self.author, ledger_info, signature }
+		return order_vote
+	}
+	return none
+}
+
+Procedure make_timeout(round, high_qc, last_tc) {
+	...
+	if valid_signature && safe_to_timeout {
+		increase_highest_vote_round(round)
+		increase_highest_timeout_round(round)
+		return TimeoutInfo
+	}
+	...
+}
+
+Procedure safe_to_order_vote(qc) {
+	// no order vote for this round if already timeout
+	return qc.vote_data.round > highest_timeout_round
+}
+
+Procedure increase_highest_timeout_round(round) {
+	highest_timeout_round <- max(highest_timeout_round, round)
+}
+```
+
+When a validator receives an order vote, it uses the following algorithm to verify and process the order vote.
+
+```
+Procedure process_order_vote_msg(order_vote_msg) {
+	// the signature of order_vote should already be verified upon receiving
+	commit_id <- order_vote_msg.order_vote.ledger_info.id
+	pending_order_votes[commit_id].add(order_vote.signature)
+	if |pending_order_votes[commit_id]| >= 2f+1 {
+	    // Add order_vote_msg.qc to the block store, if not already present
+		// Send the block with commit_id to execution
+		// Update highest_ordered_cert
+		finalize(order_vote.ledger_info)
+	}
+}
+```
+
+### Proof of Safety
```

Comment:
I think it is better to remove proofs in this AIP. They are not rigorous enough.

Reply:
@zekun000 I see you approved already. Is it okay if I remove the proofs?

Reply:
yeah I don't think it's necessary to have proofs here

Reply:
Done.